toolchain: use Kconfig to say if picolibc module can be built

dcpleung opened the pull request:
Currently, whether the picolibc module can be built is tied
to the toolchain being the Zephyr SDK. Although, currently,
only the Zephyr SDK has built-in picolibc binary, other
toolchains can use picolibc by building it from source.
To enable this behavior, a new kconfig is added:
CONFIG_TOOLCHAIN_CAN_BUILD_PICOLIBC_MODULE. This will be
enabled on a per toolchain basis to preseve the current
behavior of opt-in support for picolibc.

dcpleung committed:
libc: picolibc: disable prompt for PICOLIBC_USE_MODULE...

...when toolchain does not have native picolibc support.
Without native support, picolibc needs to be built from
the module. Disabling the prompt means this kconfig
takes on whatever default value specified, which is
to build source from module when there is no native
support.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
@keith-packard What do you think of this? This allows adding toolchain to support building picolibc without the ever increasing list of `ZEPHYR_TOOLCHAIN_VARIANT` in `CONFIG_PICOLIBC_SUPPORTED`.

keith-packard commented on cmake/toolchain/Kconfig:
```diff
@@ -0,0 +1,12 @@
+	help
+	  Toolchain has picolibc library.
+
+config TOOLCHAIN_CAN_BUILD_PICOLIBC_MODULE
```
I think the existing Kconfig bits end up assuming that all toolchains can build the picolibc module by default? Maybe this should default to 'y'?

dcpleung replied:
`CONFIG_PICOLIBC_SUPPORTED` would only allow Zephyr SDK to build the module. I am simply trying to preserve the current behavior. Since I don't have access to all toolchain, I am not comfortable in flipping the behavior mode.

keith-packard replied:
yup, I was just mistaken about what the default behavior was.

tejlmand replied:
The definition of `TOOLCHAIN_CAN_BUILD_PICOLIBC_MODULE` should be moved elsewhere.
As the setting is closely coupled to the picolibc module itself, then this could be a good location:
https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/Kconfig.picolibc

A toolchain which supports building picolibc can then do a `select TOOLCHAIN_CAN_BUILD_PICOLIBC_MODULE`, as it is done for `THREAD_LOCAL_STORAGE`.
For example:
https://github.com/zephyrproject-rtos/zephyr/blob/eee70b9d096aebc6ba8dda02efe19526ace6c369/cmake/toolchain/xtools/Kconfig#L4-L6

extra comment: would `TOOLCHAIN_SUPPORTS_PICOLIBC_MODULE` be a better name, as it aligns to other `<something>_SUPPORTS_<name>` var names used in Kconfig ?

keith-packard commented on lib/libc/Kconfig:
```diff
@@ -35,8 +35,8 @@ config NEWLIB_LIBC_SUPPORTED
 config PICOLIBC_SUPPORTED
 	bool
 	depends on !NATIVE_APPLICATION
-	depends on ("$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr") || (NATIVE_LIBRARY)
```
These conditionals are super twitchy, especially the second one. I'm unsure precisely how this works today, but for glibcxx_libcpp to work, both it and picolibc must come from the toolchain, and this condition is somehow doing that.

dcpleung replied:
Hm... `CONFIG_PICOLIBC_USE_MODULE` does `depends on !GLIBCXX_LIBCPP`. So if `GLIBCXX_LIBCPP` is enabled, building the module is not an option. Though `GLIBCXX_LIBCPP` should not be gating `CONFIG_PICOLIBC_SUPPORTED`?...

keith-packard replied:
yeah, there have been incremental changes to these configuration conditions leaving me very unsure of what the right values should be now.

keith-packard replied:
the essential requirement is that the C library and libstdc++ must both be provided by the toolchain as libstdc++ must be built atop the associated C library. The existing conditions try to limit picolibc to environments which are known to work. For C++, the only known working configuration is to have the Zephyr SDK provide both libstdc++ and picolibc; adding more combinations here when toolchains provide both picolibc and the standard C++ library.

dcpleung replied:
The kconfig combinations in this PR should keep that behavior and not allowing toolchains without `TOOLCHAIN_HAS_PICOLIBC` to use GNU C++ STL.

dcpleung committed:
cmake: expose TOOLCHAIN_HAS_PICOLIBC to kconfig

This mirrors TOOLCHAIN_HAS_NEWLIB to picolibc so the cmake
variable TOOLCHAIN_HAS_PICOLIBC can be used in kconfig.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
libc: picolibc: extend support beyond Zephyr SDK

The picolibc kconfigs were tied to ZEPHYR_TOOLCHAIN_VARIANT
and it only supported the Zephyr SDK, which is the only
toolchain having picolibc built-in at the moment. This
commit generalizes that to use TOOLCHAIN_HAS_PICOLIBC
cmake variable the same way as newlib: TOOLCHAIN_HAS_NEWLIB.
This provides the ability for other toolchains to declare
their support for picolibc.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

tejlmand commented on Kconfig.zephyr:
```diff
@@ -50,6 +50,7 @@ source "drivers/Kconfig"
 source "lib/Kconfig"
 source "subsys/Kconfig"
 
+source "cmake/toolchain/Kconfig"
```
please remove, we are including `$(TOOLCHAIN_KCONFIG_DIR)/Kconfig` right below, which allows the toolchain itself to adjust defined Kconfig.

Content of file is commented elsewhere.

dcpleung replied:
Moved.

tejlmand commented on lib/libc/picolibc/Kconfig:
```diff
@@ -5,7 +5,7 @@ if PICOLIBC
 config PICOLIBC_USE_MODULE
 	bool "Picolibc as module"
-	default y if "$(TOOLCHAIN_HAS_PICOLIBC)" != "y"
+	default y if ("$(TOOLCHAIN_HAS_PICOLIBC)" != "y") && TOOLCHAIN_CAN_BUILD_PICOLIBC_MODULE
```
this is only a default, meaning it can be changed.

But if `PICOLIBC==y` and the toolchain can only build picolibc as module, but has `TOOLCHAIN_HAS_PICOLIBC==n`, then this setting must be a forced yes.
Either by being promptless, or with a `select PICOLIBC_USE_MODULE` somewhere.

dcpleung replied:
I use `TOOLCHAIN_SUPPORTS_PICOLIBC_MODULE` to select `PICOLIBC_USE_MODULE` instead.

tejlmand reviewed (changes requested):
Generally looking fine, only some minor comments on the toolchain handling part.

dcpleung commented:
I have verified this with `cross-compile` (although with the Zephyr SDK).

tejlmand commented on lib/libc/Kconfig:
```diff
@@ -35,8 +35,8 @@ config NEWLIB_LIBC_SUPPORTED
 	depends on !NATIVE_APPLICATION
-	depends on ("$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr") || (NATIVE_LIBRARY)
-	depends on !(CPP && "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr")
+	depends on ("$(TOOLCHAIN_HAS_PICOLIBC)" = "y") || TOOLCHAIN_SUPPORTS_PICOLIBC_MODULE || NATIVE_LIBRARY
+	depends on !(CPP && (("$(TOOLCHAIN_HAS_PICOLIBC)" = "y") || TOOLCHAIN_SUPPORTS_PICOLIBC_MODULE))
```
upon second review I was looking at https://github.com/zephyrproject-rtos/sdk-ng/blob/main/cmake/zephyr/Kconfig and discovered that the reason for this code is to handle Zephyr SDK 0.15.x.

Thus we should not change this code.
If a toolchain supports picolibc (built-in or as module) then it should have the following code in its toolchain Kconfig.
```
config PICOLIBC_SUPPORTED
	default y
```
```suggestion
	depends on ("$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr") || (NATIVE_LIBRARY)
	depends on !(CPP && "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr")
```

tejlmand reviewed (changes requested):
upon second review of this PR, I discovered that we are already having the right infrastructure for this. The code related tio defaults / depends are to handle oler Zephyr SDK, like 0.15.x, SDKs still supported by Zephyr.

So what perhaps is lacking is better documentation.

For a toolchain who supports picolibc (as module or built-in), the following is required in `$(TOOLCHAIN_KCONFIG_DIR)/Kconfig`:
```
config PICOLIBC_SUPPORTED
        default y
```

For a toolchain where Picolibc is provided built-in, then the following is needed in the `${ZEPHYR_TOOLCHAIN_VARIANT}/generic.cmake`
```
set(TOOLCHAIN_HAS_PICOLIBC ON CACHE BOOL "True if toolchain supports picolibc")
```
Note, that for an oot support for the toolchain, those files are found under the `TOOLCHAIN_ROOT` setting.

Changes related to default value of `PICOLIBC_USE_MODULE` are still needed.

dcpleung commented:
Changed to use `CONFIG_PICOLIBC_SUPPORTED` instead.

tejlmand reviewed (dismissed):
changes in this PR is approved.

@dcpleung are you planning on improving Zephyr docs in addition to this PR ?
Perhaps on this page: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/languages/c/picolibc.rst

so that other's facing the same issue as you have experienced knows how to correctly specify that toolchain supports picolibc.

dcpleung commented:
> @dcpleung are you planning on improving Zephyr docs in addition to this PR ? Perhaps on this page: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/languages/c/picolibc.rst

Let me add the info to doc.

dcpleung committed:
doc: picolibc: a few words about building without native support

This adds a few words about using Picolibc without native support
from toolchain, and how to enable this behavior for toolchain.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Add a few words about toolchain without native picolibc support.

carlescufi commented:
@keith-packard are you OK with this?
